### ISSUE TITLE
Implement visual reports

### ISSUE THREAD

**Issue by emdupre** [nilearn/nilearn#2022]
PR: #2019.

Generating visual reports may help users to more closely understand what is occurring during processing.

Example use case: A common error with NiftiMasker is that the mask is not aligned to the reference image. Extracted time series therefore do not correspond to intended functional areas and may lead to puzzling results. If users are able to visualize the alignment of the mask with the reference image, they can quickly identify the problem.

The initial structure will be quite simple: A report title (based on the function it's called from, and potentially the file(s) called on), a visualization, and an optional description of the visualization, indicating how it should be used (e.g., things to watch out for). 

My current plan is to try and write a super class, but only implement this for NiftiMasker in a first pass. To do so, we'll add:

1. A new HTML report template
2. A vendored copy of _tempita_ to use for templating
3. The actual functionality

Very much looking for feedback ! Any and all comments appreciated :cherry_blossom:

**Comment by emdupre** [comment]
Pinging @GaelVaroquaux @jeromedockes to continue the discussion

**Comment by GaelVaroquaux** [comment]
The exact specification of what we put in a report will evolve with time and will depend on the object. The idea is to start simple and prototype with the maskers, and then move on to object like CanICA.

One thing that we will need to worry about in the long run is how to compose reports: an object like CanICA should embed the report of the NiftiMasker it uses.

We will need to focus on good UX (user experience). For this, two possible tricks:

- Tooltips that appear on hover
- Parts of the report that is hidden by default.

Pure css / html implementation (no javascript) can be found on my website http://gael-varoquaux.info/about.html
- Hovering on the portrait displays a tooltip
- The email is hidden by an "email" button.

**Comment by emdupre** [comment]
> We will need to focus on good UX (user experience). 

Other long-term ideas currently under discussion include a show/hide code button, maybe building on the implementation we currently have in [Jupyter Book](https://jupyter.org/jupyter-book/features/hiding.html).

**Comment by GaelVaroquaux** [comment]
To embed an image in an html:

"""
from matplotlib import pyplot as plt
import io
import base64

io_buffer = io.BytesIO()
plt.figure()
plt.axes()
plt.savefig(io_buffer, format='png')

io_buffer.seek(0)

data = base64.b64encode(io_buffer.read())

html_str = '''
    <html>
    <img src="data:image/png;base64,{0}" alt="beastie.png" scale="0">
    </html>
    ''.format(data.decode())

with open("tmp.html", "w") as outfile:
    outfile.write(html_str)


"""

**Comment by GaelVaroquaux** [comment]
To have tooltips in pure CSS:

CSS code:
```
[tooltip]:hover:before {
 background:rgba(0,0,0,.8);
 border-radius:5px;
 color:#fff;
 content:attr(tooltip);
 left:50%;
 padding:5px 15px;
 position:absolute;
 z-index:98;
 width:50px;
 margin-left:-40px;
 bottom:100%;
 display:inline-block;
 text-align:center
}
```

HTML code:
```
<a tooltip="A tooltip">A link</a>
```

**Comment by emdupre** [comment]
Right now reports are being generated using something like the following:

```python
from nilearn.input_data import NiftiMasker

# NiftiMasker with an example file from the brain development dataset
f = 'sub-pixar001_task-pixar_space-MNI152NLin2009cAsym_desc-preproc_bold.nii.gz'
mask = NiftiMasker(verbose=True)
mask.fit(f)
html = mask.generate_report()
html.open_in_browser()
```

The reports also display in the Jupyter Notebook. LMK if you see any concerns, there !

**Comment by emdupre** [comment]
Dropping `ReportMixin`, the new call is:

```python
from nilearn.input_data import NiftiMasker
from nilearn.plotting import generate_report

# NiftiMasker with an example file from the brain development dataset
f = 'sub-pixar001_task-pixar_space-MNI152NLin2009cAsym_desc-preproc_bold.nii.gz'
mask = NiftiMasker(verbose=True)
mask.fit(f)
html = generate_report(mask)
html.open_in_browser()
```

**Comment by GaelVaroquaux** [comment]
> Dropping ReportMixin, the new call is:

> html = generate_report(mask)

Right. But I would like to go back to the method. From a UX it seems
better. From a code design, OOP is also desirable.

The problem of isolation can be tackled with safe plotting functions, as
mentioned in
https://github.com/nilearn/nilearn/pull/2019#issuecomment-488667470

**Comment by emdupre** [comment]
The current problem on CircleCI is that when we call `.fit_transform()`, the `.fit` internal call isn't being passed any `imgs`. So we're getting a `TypeError` because this snippet:

```python
# save inputs for reporting
dim = image.load_img(X).shape
img = image.index_img(X, dim[-1] // 2)
self.input_ = img
```

is being called on `None`. The issue is that I'm having trouble finding a way to consistently save the inputs to `NiftiMasker`, given these multiple options for calling. Without saving the input, somewhere, we can't generate plots for the report -- at least in my current thinking.

Any suggestions here ?

**Comment by kchawla-pi** [comment]
What's stopping us from saving the inputs somewhere?

**Comment by jeromedockes** [comment]
> is being called on `None`. The issue is that I'm having trouble finding a way to consistently save the inputs to `NiftiMasker`, given these multiple options for calling. Without saving the input, somewhere, we can't generate plots for the report -- at least in my current thinking.

if a mask image was provided, this is used to mask data and the `fit` method doesn't compute a new mask. in this case I guess there is not much to report. you could store `None`  as the input to `fit`, and in the report say that the mask image was provided in the `__init__` and possibly display it.

**Comment by jeromedockes** [comment]
> DIT: This general case of `.fit_transform()` presents several other problems, to be clear. The biggest is that we don't save the `NiftiMasker` class out, and only keep the `np.ndarray` from the `.transform`, which doesn't have all the attributes we need to access when creating reports.

`fit_transform` will still modify the transformer (ie `NiftiMasker`) in place, so if you store attributes in `fit` they will be there. in other words
```
masker = NiftiMasker()
masked = masker.fit_transform(images) # masker is being modified
generate_report(masker)
```
should be fine. of couse if a user doesn't have a reference to the masker object itself they cannot create a report

**Comment by jeromedockes** [comment]
Also, I suggest we post future comment in the PR's discussion, to have it all in one place

**Comment by kchawla-pi** [comment]
Wait you guys have been talking shop behind our backs 😱😀

**Comment by jeromedockes** [comment]
> Wait you guys have been talking shop behind our backs screamgrinning

nope, I mean that the discussion is split between this issue and the PR #2019

**Comment by emdupre** [comment]
> Also, I suggest we post future comment in the PR's discussion, to have it all in one place

I think I still have some general confusion around how nilearn uses issues vs PRs. Should I just close this issue, then, to have all the discussion focused in the PR ?

**Comment by kchawla-pi** [comment]
Nah don't close the issue.

Issues are filled to point out a problem or feature and discuss it until a
PR is opened. The implementation details are discussed in the PR.
Once the PR is merged the issue is closed.
